server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -202,4 +202,75 @@ public Releasable dropHeadersContext() {
             return dropHeadersContextSupplier.get();
         }
     }
+
+    /**
+     * A basic implementation for batch executors that simply need to execute the tasks in the batch iteratively,
+     * producing a cluster state after each task. This allows executing the tasks in the batch as a
+     * series of executions, each taking an input cluster state and producing a new cluster state that serves as the
+     * input of the next task in the batch.
+     */
+    abstract class DefaultBatchExecutor<T extends ClusterStateTaskListener> implements ClusterStateTaskExecutor<T> {
```

Comment:
Could we make `void clusterStatePublished(ClusterState)` a `final` method and allow an override `void clusterStatePublished()` which doesn't expose the published state? Implementations shouldn't be using the published state I think.